ext/java/org/jruby/ext/stringio/StringIO.java

```diff
@@ -1636,6 +1649,73 @@ public IRubyObject set_encoding(ThreadContext context, IRubyObject enc, IRubyObj
         return set_encoding(context, enc);
     }
 
+    @JRubyMethod
+    public IRubyObject set_encoding_by_bom(ThreadContext context) {
+        if (setEncodingByBOM(context) == null) return context.nil;
+
+        return context.runtime.getEncodingService().convertEncodingToRubyEncoding(ptr.enc);
+    }
+
+    private Encoding setEncodingByBOM(ThreadContext context) {
+        Encoding enc = detectBOM(context, ptr.string, (ctx, enc2, bomlen) -> {
+            ptr.pos = bomlen;
+            if (writable()) {
+                ptr.string.setEncoding(enc2);
+            }
+            return enc2;
+        });
+        ptr.enc = enc;
+        return enc;
+    }
+
+    private static Encoding detectBOM(ThreadContext context, RubyString str, ObjectObjectIntFunction<ThreadContext, Encoding, Encoding> callback) {
+        int p;
+        int len;
+
+        ByteList byteList = str.getByteList();
+        byte[] bytes = byteList.unsafeBytes();
+        p = byteList.begin();
+        len = byteList.realSize();
+
+        if (len < 1) return null;
+        switch (toUnsignedInt(bytes[p])) {
+            case 0xEF:
+                if (len < 2) break;
+                if (toUnsignedInt(bytes[p + 1]) == 0xBB && len > 2) {
```

Comment:
Just a nitpick: why `len < 2`?

```suggestion
                if (len < 3) break;
                if (toUnsignedInt(bytes[p + 1]) == 0xBB) {
```

Reply:
The original logic was ported from the C code:

https://github.com/ruby/stringio/blob/master/ext/stringio/stringio.c#L229-L230

Are you sure this should change to `len < 3`?

Reply:
I believe this is resolved in the other comment from @kou.